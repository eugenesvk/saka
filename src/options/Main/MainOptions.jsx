import { Component, h } from 'preact';
import 'material-components-web/dist/material-components-web.css';
import 'scss/options.scss';

import OptionsList from './OptionsList/index.jsx';
import SakaHotkeysList from './SakaHotkeysList/index.jsx';

export default class MainOptions extends Component {
  constructor(props) {
    super(props);

    this.state = {
      showSakaKeybindings: false
    };
  }

  handleOpenSakaKeybindings = () => {
    this.setState({
      showSakaKeybindings: !this.state.showSakaKeybindings
    });
  };

  render() {
    return (
      <body>
        <header className="mdc-top-app-bar mdc-top-app-bar--short">
          <div className="mdc-top-app-bar__row">
            <section className="mdc-top-app-bar__section mdc-top-app-bar__section--align-start">
              <span className="mdc-top-app-bar__title">Saka Options</span>
            </section>
          </div>
        </header>
        <div
          id="background-image"
          className="mdc-elevation--z1 options-container"
        >
          {this.state.showSakaKeybindings ? (
            <SakaHotkeysList
              handleOpenSakaKeybindings={this.handleOpenSakaKeybindings}
            />
          ) : (
            <OptionsList
              handleOpenSakaKeybindings={this.handleOpenSakaKeybindings}
            />
          )}
        </div>
      </body>
    );
  }
}
