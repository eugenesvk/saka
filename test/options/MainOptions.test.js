import { h } from 'preact';
import {
  render,
  cleanup,
  wait,
  fireEvent,
  flushPromises,
  getByValue
} from '@testing-library/preact';
import '@testing-library/jest-dom/extend-expect';
import MainOptions from '@/options/Main/MainOptions.jsx';

beforeEach(() => {
  browser.flush();
  browser.storage.sync.set.returns({});
});

test('should show all options when not showing key bindings', async () => {
  browser.storage.sync.get.returns({});
  const { getByText, queryByText } = render(<MainOptions />);

  getByText('Saka Options');
  await wait(() => getByText('General Settings'));
  //DefaultModeSelection
  getByText('Default Mode');
  getByText('Select the default mode Saka opens with');
  //OnlyShowSearchBarSelector
  getByText('Suggestions on load');
  getByText('Show suggestions when there is no text is the Saka search bar');
  //EnableFuzzySearch
  getByText('Enable fuzzy search');
  getByText('Enable fuzzy search for bookmarks and history search');

  expect(queryByText('Keyboard Shortcuts')).not.toBeInTheDocument();
});

test('should only show key bindings when setting is true', async () => {
  browser.storage.sync.get.returns({});
  global.SAKA_PLATFORM = 'chrome';
  const { debug, getByText, getByLabelText } = render(<MainOptions />);

  await wait(() => getByText('General Settings'));
  getByText('Saka Hotkeys');
  fireEvent.click(getByText('keyboard'), { button: 0 });
  await flushPromises();

  getByText('Saka Options');
  expect(getByText('arrow_back'));
  expect(getByLabelText('Back to Saka settings')).toMatchSnapshot();
  expect(getByLabelText('Info about Saka custom hotkeys')).toMatchSnapshot();
  getByText(
    'To modify the Saka hotkeys, please visit chrome://extensions/shortcuts'
  );
  getByText('Keyboard Shortcuts');
  getByText('Open Saka');
});

test('should save settings when save button clicked', async () => {
  browser.storage.sync.get.returns({
    sakaSettings: {}
  });
  const { getByText, getByLabelText, getByValue } = render(<MainOptions />);

  await wait(() => getByText('General Settings'));

  fireEvent.change(getByLabelText('Select default mode'), {
    target: { value: 'history' }
  });
  await flushPromises();

  fireEvent.click(getByLabelText('Suggestions on load'), { button: 0 });
  await flushPromises();

  fireEvent.click(getByLabelText('Enable fuzzy search'), { button: 0 });
  await flushPromises();

  fireEvent.click(getByValue('Save'), { button: 0 });
  await flushPromises();

  expect(browser.storage.sync.get.calledOnce);
});

afterEach(cleanup);
