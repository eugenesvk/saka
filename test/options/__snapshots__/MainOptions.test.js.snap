// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should only show key bindings when setting is true 1`] = `
<i
  aria-label="Back to Saka settings"
  aria-pressed="false"
  class="mdc-icon-button material-icons"
  role="button"
>
  arrow_back
</i>
`;

exports[`should only show key bindings when setting is true 2`] = `
<i
  aria-label="Info about Saka custom hotkeys"
  aria-pressed="false"
  class="mdc-icon-button material-icons"
  id="custom-hotkey-info"
>
  info
</i>
`;
